distribution/metadata/v2_metadata_service.go

```diff
@@ -15,10 +18,24 @@ type V2MetadataService struct {
 
 // V2Metadata contains the digest and source repository information for a layer.
 type V2Metadata struct {
+	AuthConfigHash   string
 	Digest           digest.Digest
 	SourceRepository string
 }
 
+// HashAuthConfig returns an sha256 sum of the "authConfig" if successful.
+func HashAuthConfig(authConfig *types.AuthConfig) (string, error) {
+	if authConfig == nil {
+		return "", nil
+	}
+	buf, err := json.Marshal(authConfig)
```

Comment:
I think I'm okay with including the password in the hash input, but want to take care that we handle anything including passwords sensitively. One suggestion would be to store a HMAC of the record data using a hash of `AuthConfig` as the key. For example:

```go
	authConfigDigest := digest.FromBytes([]byte(hashInput))
	mac := hmac.New(sha256.New, []byte(authConfigDigest))
	mac.Write([]byte(v2Metadata.LayerDigest))
	mac.Write([]byte(v2Metadata.SourceRepository))
	expectedMAC := mac.Sum(nil)
	if hmac.Equal(v2Metadata.StoredMAC, expectedMAC) {
		// This AuthConfig is the same one the record was created with
	}
```

In this approach, the other fields of `V2Metadata` are covered by the HMAC, so they effectively provide a salt for the sensitive information we're hashing from the `AuthConfig`. It would be very difficult for someone to extract useful information about `AuthConfig` from the result because the details of the record are hashed in. It would prevent attacks where someone checks hashes against a table of hash results pregenerated from known username/password pairs, for example.